src/haz3lcore/dynamics/DH.re

```diff
@@ -1,52 +1,55 @@
 open Sexplib.Std;
+open Hazel_menhir.AST;
 
 [@deriving (show({with_path: false}), sexp, yojson)]
 type if_consistency =
   | ConsistentIf
   | InconsistentIf;
 
 module rec DHExp: {
+  //All comments show the textual syntax for DHExp when built using the menhir parser
   [@deriving (show({with_path: false}), sexp, yojson)]
   type t =
-    | EmptyHole(MetaVar.t, HoleInstanceId.t)
-    | NonEmptyHole(ErrStatus.HoleReason.t, MetaVar.t, HoleInstanceId.t, t)
-    | FreeVar(MetaVar.t, HoleInstanceId.t, Var.t)
+    | EmptyHole(MetaVar.t, HoleInstanceId.t) //()
+    | NonEmptyHole(ErrStatus.HoleReason.t, MetaVar.t, HoleInstanceId.t, t) //(_HOLE x)
+    | FreeVar(MetaVar.t, HoleInstanceId.t, Var.t) //(_FREE x)
     | InvalidText(MetaVar.t, HoleInstanceId.t, string)
-    | InconsistentBranches(MetaVar.t, HoleInstanceId.t, case)
+    | InconsistentBranches(MetaVar.t, HoleInstanceId.t, case) //(_HOLE (case ...))
     | Closure([@opaque] ClosureEnvironment.t, t)
-    | Filter(DHFilter.t, t)
-    | BoundVar(Var.t)
-    | Sequence(t, t)
-    | Let(DHPat.t, t, t)
-    | FixF(Var.t, Typ.t, t)
-    | Fun(DHPat.t, Typ.t, t, option(Var.t))
+    | Filter(DHFilter.t, t) //pause a == 2 b
+    | BoundVar(Var.t) //x
+    | Sequence(t, t) //a; b
+    | Let(DHPat.t, t, t) //let x = y in z
+    | FixF(Var.t, Typ.t, t) //_FIX f Int -> Int fun: Int x -> 1 + x f in 55
+    | Fun(DHPat.t, Typ.t, t, option(Var.t)) //fun: Int x -> x + 1
```

Comment:
the syntax here should be `fun p -> e`

Reply:
I had to add in manual annotations for the type of the function to be able to construct the dhexp with the type. The current elaborator gets the type from the pattern but uses the statics map - I thought it would be best to just add in the annotations rather than try to interface with that. What do you think?